### PR TITLE
Bump up data table version to V4 to optimize space usage.

### PR DESCRIPTION
**Pull request by mqliang**
## Description

ref: https://github.com/apache/incubator-pinot/issues/6726

This PR bumps up data table version to V4 to address TODOs in DataTableBuilder:

 * Fix float size: V2/V3 allocate 8 bytes to store float, while only use 4 bytes. V4 allocate and use 4 bytes.
 * V2/V3 use one dictionary for each columns to map String->Integer, while V4 use a common dictionary for all columns.
 * V4 store bytes as variable size data instead of String.

In terms of the TODO of "Given a data schema, write all values one by one instead of using rowId and colId to position (save time)." Based on the benchmark of https://github.com/apache/incubator-pinot/issues/6720, it's over-engineering thus not  needed. 

I also use the compatibility-verifier tool to verify the compatibility locally, this PR does not break the back-comp.

## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)

### CODE REVIEW

**Comment by xiangfu0** [comment]
since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use? 
So the Pinot server can work with v3 broker/v4 broker or v2 presto.

**Comment by mqliang** [comment]
test failed due to flaky issue. close and re-open to trigger a re-run

**Comment by mayankshriv** [comment]
> since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use?
> So the Pinot server can work with v3 broker/v4 broker or v2 presto.

@mqliang I have the same concern as @xiangfu0. We already ran into this issue when you did the v2 to v3 upgrade. Even though this might be thought of as internal Pinot data structure, there are connectors that use this data table, making it an externally exposed data structure, and we need to be careful with incrementing version numbers in backward incompatible way.

**Comment by mqliang** [comment]
> since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use?
So the Pinot server can work with v3 broker/v4 broker or v2 presto.

@xiangfu0 
We will eventually stop supporting V2 and V3, so it's better to update codebase of presto connector to follow up the most recent pinot. But I do think it a good idea to to allow client specify DataTable version, this will make client happy: client only need make minimal change after pinot is upgraded( by specifying a old DataTable version), this will give client developer more time to upgrade their code. Let's pull @mcvsubbu @siddharthteotia to chime in.

**Comment by mayankshriv** [comment]
> > since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use?
> > So the Pinot server can work with v3 broker/v4 broker or v2 presto.
> 
> @xiangfu0 @mayankshriv
> We will eventually stop supporting V2 and V3, so it's better to update codebase of presto connector to follow up the most recent pinot. But I do think it a good idea to to allow client specify DataTable version, this will make client happy: client only need make minimal change after pinot is upgraded( by specifying an old DataTable version), this will give client developer more time to upgrade their code. Let's pull @mcvsubbu @siddharthteotia to chime in.

@mqliang We do not own the Presto code. On top of that, there are two of them PrestoDB and Trino. And is being used in production at Uber and City Storage Systems. Also, we are still dealing with the v3 change that was recently made.

FWIW, I thikn either DataTable should not have been exposed outside of Pinot, or it should have been treated as an external interface that should not be changed in backward incompatible way. But given that we are in the situation, we should probably acknowledge that and avoid making further backward incompatible changes until we resolve the situation across the communities.

Thoughts?

**Comment by mayankshriv** [comment]
Adding @yupeng9 and @elonazoulay for visibility.

**Comment by xiangfu0** [comment]
> > since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use?
> > So the Pinot server can work with v3 broker/v4 broker or v2 presto.
> 
> @mqliang I have the same concern as @xiangfu0. We already ran into this issue when you did the v2 to v3 upgrade. Even though this might be thought of as internal Pinot data structure, there are connectors that use this data table, making it an externally exposed data structure, and we need to be careful with incrementing version numbers in backward incompatible way.

Agreed. I think the current blocker is that both v3/v4 are not part of the 0.7.1 release, so there is no way for external dependencies to upgrade. Ideally, this feature should be checked in for 1 or 2 releases then we can make the switch.

**Comment by mcvsubbu** [comment]
> > > since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use?
> > > So the Pinot server can work with v3 broker/v4 broker or v2 presto.
> > 
> > 
> > @xiangfu0 @mayankshriv
> > We will eventually stop supporting V2 and V3, so it's better to update codebase of presto connector to follow up the most recent pinot. But I do think it a good idea to to allow client specify DataTable version, this will make client happy: client only need make minimal change after pinot is upgraded( by specifying an old DataTable version), this will give client developer more time to upgrade their code. Let's pull @mcvsubbu @siddharthteotia to chime in.
> 
> @mqliang We do not own the Presto code. On top of that, there are two of them PrestoDB and Trino. And is being used in production at Uber and City Storage Systems. Also, we are still dealing with the v3 change that was recently made.
> 
> FWIW, I think either DataTable should not have been exposed outside of Pinot, or it should have been treated as an external interface that should not be changed in backward incompatible way. But given that we are in the situation, we should probably acknowledge that and avoid making further backward incompatible changes until we resolve the situation across the communities.
> 
> Thoughts?

Totally agree that we should not have exposed DataTable to begin with.  I believe we do need this optimization, so adding something in the protocol (or, maybe adding a capability handshake in the beginning?) is workable.  Note that adding this in the beginning will mean adding a config to the broker to go ahead and use v2,v3,v4 in its requests. Since we upgrade brokers first, this will also mean  for the server to look (for each request) which version it needs to return with, and exception if it does not support that version (for a robust protocol design). It is also possible for the server to indicate its support in zookeeper under its instance config, and the broker can cache it on a per server basis along with the routing table. Either handshake or zk seems to be a workable solution, probably better than adding something in every query path.

Can we also consider adding an spi interface for datatable, and have the trino/presto community make a one-time move to use that instead? Over time, we can support the concept of a session, and retain such things at session level rather than every query.

**Comment by mcvsubbu** [comment]
FWIW. I was going to suggest that we deprecate V2 in this commit (and remove support for it in 0.9+)

**Comment by dharakk** [comment]
> > > since presto connector will access pinot server with segment query and wrap up DataTable by itself. Is it possible to make the query itself contain the version of which DataTable protocol to use?
> > > So the Pinot server can work with v3 broker/v4 broker or v2 presto.
> > 
> > 
> > @xiangfu0 @mayankshriv
> > We will eventually stop supporting V2 and V3, so it's better to update codebase of presto connector to follow up the most recent pinot. But I do think it a good idea to to allow client specify DataTable version, this will make client happy: client only need make minimal change after pinot is upgraded( by specifying an old DataTable version), this will give client developer more time to upgrade their code. Let's pull @mcvsubbu @siddharthteotia to chime in.
> 
> @mqliang We do not own the Presto code. On top of that, there are two of them PrestoDB and Trino. And is being used in production at Uber and City Storage Systems. Also, we are still dealing with the v3 change that was recently made.
> 
> FWIW, I think either DataTable should not have been exposed outside of Pinot, or it should have been treated as an external interface that should not be changed in backward incompatible way. But given that we are in the situation, we should probably acknowledge that and avoid making further backward incompatible changes until we resolve the situation across the communities.
> 
> Thoughts?

Hi +1 to this! Uber has use cases using the direct segment query path via Presto and backward incompatible change will be troublesome. Would be great to have the protocol to specify the version first and then the datatable upgrade in the following releases.

**Comment by mqliang** [comment]
test failed due to flaky issue. close and re-open to trigger a re-run

File: pinot-core/src/main/java/org/apache/pinot/core/common/datatable/BaseDataTable.java [hunk 1]
```diff
@@ -44,21 +44,24 @@
   protected DataSchema _dataSchema;
   protected int[] _columnOffsets;
   protected int _rowSizeInBytes;
+  // _dictionaryMap is only used in V2/V3, where each column use one dictionary to map String->Integer
   protected Map<String, Map<Integer, String>> _dictionaryMap;
+  // _dictionary is only used in V4, where all columns use a common dictionary to map String->Integer
+  protected Map<Integer, String> _dictionary;
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/common/datatable/BaseDataTable.java:50]: We can further optimize this into a `List`

Comment (mqliang) [reply pinot-core/src/main/java/org/apache/pinot/core/common/datatable/BaseDataTable.java:50]: That's a good idea, using the index of list as an implicit dicId.

**Comment by mqliang** [comment]
@mayankshriv @dharakk

Should we keep back-comp of DataTable at interface level or at binary level (the bytes layout when transfer on wire)? 

* If presto connector import pinot code as a package to build a DataTable or to parse a bytes as a DataTable, then we only only need interface level back-comp. Then we will be fine -- this change does not change the DataTable interface. The presto connector just need to bump up its pinot dependency version. 
* If presto connector manipulate DataTable bytes by itself, we then need binary level back-comp. The the situation will becomes much more complex -- we need a protocol:
     - between pinot-server and pinot-broker
     - between pinot-server and presto-connector
     - between pinot-broker and presto-connector


@xiangfu0 
>  Is it possible to make the query itself contain the version of which DataTable protocol to use?
So the Pinot server can work with v3 broker/v4 broker or v2 presto.

Instead of specifying the version for each query, we can change the server config of `pinot.server.instance.currentDataTableVersion`. In this way, new server will continue to return old version DataTable, old broker/presto-connector can continue work.

**Pull request closed without merging**